File: features/digital/flight/src/main/java/com/tokopedia/flight/searchV2/presentation/activity/FlightSearchReturnActivity.java
package com.tokopedia.flight.searchV2.presentation.activity;

import android.content.Context;
import android.content.Intent;
import android.support.v4.app.Fragment;

import com.tokopedia.flight.airport.view.viewmodel.FlightAirportViewModel;
import com.tokopedia.flight.booking.view.activity.FlightBookingActivity;
import com.tokopedia.flight.common.constant.FlightFlowExtraConstant;
import com.tokopedia.flight.common.util.FlightAnalytics;
import com.tokopedia.flight.common.util.FlightDateUtil;
import com.tokopedia.flight.common.util.FlightFlowUtil;
import com.tokopedia.flight.search.view.model.FlightSearchPassDataViewModel;
import com.tokopedia.flight.searchV2.presentation.fragment.FlightSearchFragment;
import com.tokopedia.flight.searchV2.presentation.fragment.FlightSearchReturnFragment;

public class FlightSearchReturnActivity extends FlightSearchActivity
        implements FlightSearchFragment.OnFlightSearchFragmentListener {

    public static final String EXTRA_DEPARTURE_ID = "EXTRA_DEPARTURE_ID";
    private static final int REQUEST_CODE_BOOKING = 13;

    private String selectedDepartureID;

    public static Intent getCallingIntent(Context context,
                                          FlightSearchPassDataViewModel passDataViewModel,
                                          String selectedDepartureID) {
        Intent intent = new Intent(context, FlightSearchReturnActivity.class);
        intent.putExtra(EXTRA_PASS_DATA, passDataViewModel);
        intent.putExtra(EXTRA_DEPARTURE_ID, selectedDepartureID);
        return intent;
    }

    @Override
    protected void initializeToolbarData() {
        selectedDepartureID = getIntent().getStringExtra(EXTRA_DEPARTURE_ID);

        dateString = FlightDateUtil.formatDate(
                FlightDateUtil.DEFAULT_FORMAT,
                FlightDateUtil.DEFAULT_VIEW_FORMAT,
                passDataViewModel.getReturnDate()
        );
        passengerString = buildPassengerTextFormatted(passDataViewModel.getFlightPassengerViewModel());
        classString = passDataViewModel.getFlightClass().getTitle();
    }

    @Override
    protected FlightAirportViewModel getDepartureAirport() {
        return passDataViewModel.getArrivalAirport();
    }

    @Override
    protected Fragment getNewFragment() {
        return FlightSearchReturnFragment.newInstance(passDataViewModel, selectedDepartureID);
    }

    @Override
    protected FlightAirportViewModel getArrivalAirport() {
        return passDataViewModel.getDepartureAirport();
    }

    @Override
    public void selectFlight(String selectedFlightID) {
        // TODO : PASS FLIGHTPRICEVIEWMODEL
        startActivityForResult(FlightBookingActivity
                        .getCallingIntent(this,
                                passDataViewModel,
                                selectedDepartureID,
                                selectedFlightID,
                                null),
                REQUEST_CODE_BOOKING);
    }

    @Override
    public String getScreenName() {
        return FlightAnalytics.Screen.SEARCH_RETURN;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case REQUEST_CODE_BOOKING:
                if (data != null) {
                    FlightFlowUtil.actionSetResultAndClose(this,
                            getIntent(),
                            data.getIntExtra(FlightFlowExtraConstant.EXTRA_FLOW_DATA, 0)
                    );
                }
                break;
        }
    }
}

File: features/travel/flight/src/main/java/com/tokopedia/flight/search/presentation/activity/FlightSearchReturnActivity.kt
package com.tokopedia.flight.search.presentation.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import com.tokopedia.flight.airport.view.model.FlightAirportModel
import com.tokopedia.flight.bookingV3.presentation.activity.FlightBookingActivity
import com.tokopedia.flight.common.constant.FlightFlowExtraConstant
import com.tokopedia.flight.common.util.FlightAnalytics
import com.tokopedia.flight.common.util.FlightDateUtil
import com.tokopedia.flight.common.util.FlightFlowUtil
import com.tokopedia.flight.search.presentation.fragment.FlightSearchFragment
import com.tokopedia.flight.search.presentation.fragment.FlightSearchReturnFragment
import com.tokopedia.flight.search.presentation.model.FlightPriceModel
import com.tokopedia.flight.search.presentation.model.FlightSearchPassDataModel

/**
 * @author by furqan on 15/04/2020
 */
class FlightSearchReturnActivity : FlightSearchActivity(),
        FlightSearchFragment.OnFlightSearchFragmentListener {

    private var selectedDepartureId: String = ""
    private var selectedDepartureTerm: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        selectedDepartureId = intent.getStringExtra(EXTRA_DEPARTURE_ID)
        selectedDepartureTerm = intent.getStringExtra(EXTRA_DEPARTURE_TERM)

        super.onCreate(savedInstanceState)
    }

    override fun getNewFragment(): Fragment? {
        val priceModel: FlightPriceModel = intent.getParcelableExtra(EXTRA_PRICE_MODEL)
        return FlightSearchReturnFragment.newInstance(flightSearchPassDataModel,
                selectedDepartureId,
                intent.getBooleanExtra(EXTRA_IS_BEST_PAIRING, false),
                priceModel,
                intent.getBooleanExtra(EXTRA_IS_COMBINE_DONE, true))
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            REQUEST_CODE_BOOKING -> {
                if (data != null) {
                    FlightFlowUtil.actionSetResultAndClose(this, intent,
                            data.getIntExtra(FlightFlowExtraConstant.EXTRA_FLOW_DATA, 0))
                }
            }
        }
    }

    override fun getScreenName(): String = FlightAnalytics.Screen.SEARCH_RETURN

    override fun getDepartureAirport(): FlightAirportModel = flightSearchPassDataModel.arrivalAirport

    override fun getArrivalAirport(): FlightAirportModel = flightSearchPassDataModel.departureAirport

    override fun selectFlight(selectedFlightID: String, selectedTerm: String, flightPriceModel: FlightPriceModel, isBestPairing: Boolean, isCombineDone: Boolean, requestId: String) {
        flightSearchPassDataModel.searchRequestId = requestId
        startActivityForResult(FlightBookingActivity.getCallingIntent(this,
                flightSearchPassDataModel, selectedDepartureId, selectedDepartureTerm,
                flightPriceModel, selectedFlightID, selectedTerm), REQUEST_CODE_BOOKING)
    }

    override fun isReturnPage(): Boolean = true

    override fun initializeToolbarData() {
        dateString = FlightDateUtil.formatDate(
                FlightDateUtil.DEFAULT_FORMAT,
                FlightDateUtil.FORMAT_DATE_SHORT_MONTH,
                flightSearchPassDataModel.returnDate)
        passengerString = buildPassengerTextFormatted(flightSearchPassDataModel.flightPassengerModel)
        classString = flightSearchPassDataModel.flightClass.title
    }

    companion object {
        const val EXTRA_DEPARTURE_ID = "EXTRA_DEPARTURE_ID"
        const val EXTRA_DEPARTURE_TERM = "EXTRA_DEPARTURE_TERM"
        const val EXTRA_IS_BEST_PAIRING = "EXTRA_IS_BEST_PAIRING"
        const val EXTRA_PRICE_MODEL = "EXTRA_PRICE_MODEL"
        const val EXTRA_IS_COMBINE_DONE = "EXTRA_IS_COMBINE_DONE"

        private const val REQUEST_CODE_BOOKING = 13

        fun getCallingIntent(context: Context,
                             passDataModel: FlightSearchPassDataModel,
                             selectedDepartureId: String,
                             selectedDepartureTerm: String,
                             isBestPairing: Boolean,
                             priceModel: FlightPriceModel,
                             isCombineDone: Boolean): Intent =
                Intent(context, FlightSearchReturnActivity::class.java)
                        .putExtra(EXTRA_PASS_DATA, passDataModel)
                        .putExtra(EXTRA_DEPARTURE_ID, selectedDepartureId)
                        .putExtra(EXTRA_DEPARTURE_TERM, selectedDepartureTerm)
                        .putExtra(EXTRA_IS_BEST_PAIRING, isBestPairing)
                        .putExtra(EXTRA_PRICE_MODEL, priceModel)
                        .putExtra(EXTRA_IS_COMBINE_DONE, isCombineDone)
    }
}